File: python/canopy_ai/discover.py
"""Service discovery: GET /api/services.

The shape-translation logic lives here; the sync and async clients each call
their own request method around it.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import urlencode

if TYPE_CHECKING:
    from canopy_ai.transport import Transport
    from canopy_ai.types import DiscoverArgs, DiscoveredService


def build_query(agent_id: str | None, args: "DiscoverArgs") -> str:
    """Render a /api/services query string from discover args + agent id."""
    pairs: list[tuple[str, str]] = []
    cats = args.get("category")
    if cats is not None:
        if isinstance(cats, str):
            pairs.append(("category", cats))
        else:
            pairs.extend(("category", c) for c in cats)
    if (q := args.get("query")):
        pairs.append(("q", q))
    if args.get("include_unverified"):
        pairs.append(("include_unverified", "true"))
    if args.get("include_blocked"):
        pairs.append(("include_blocked", "true"))
    if (limit := args.get("limit")) is not None:
        pairs.append(("limit", str(limit)))
    if agent_id:
        pairs.append(("agent_id", agent_id))
    return urlencode(pairs)


def map_response(body: Any) -> list["DiscoveredService"]:
    """Convert the JSON response body's services into snake_case TypedDicts."""
    assert isinstance(body, dict), "discover response must be an object"
    services = body.get("services") or []
    return [
        {
            "slug": s.get("slug", ""),
            "name": s.get("name", ""),
            "description": s.get("description"),
            "category": s.get("category", ""),
            "logo_url": s.get("logoUrl"),
            "docs_url": s.get("docsUrl"),
            "payment_methods": [
                {
                    "realm": pm.get("realm", ""),
                    "base_url": pm.get("baseUrl", ""),
                    "protocol": pm.get("protocol", ""),
                }
                for pm in (s.get("paymentMethods") or [])
            ],
            "endpoints": [
                {
                    "method": ep.get("method", ""),
                    "path": ep.get("path", ""),
                    "description": ep.get("description"),
                    "price_atomic": ep.get("priceAtomic"),
                    "currency": ep.get("currency"),
                    "pricing_model": ep.get("pricingModel"),
                    "protocol": ep.get("protocol"),
                }
                for ep in (s.get("endpoints") or [])
            ],
            "preferred_base_url": s.get("preferredBaseUrl"),
            "policy_allowed": bool(s.get("policyAllowed", True)),
        }
        for s in services
    ]


def discover(
    transport: "Transport",
    agent_id: str | None,
    args: "DiscoverArgs",
) -> list["DiscoveredService"]:
    qs = build_query(agent_id, args)
    path = f"/api/services?{qs}" if qs else "/api/services"
    _, body = transport.request("GET", path, expect_statuses=[200])
    return map_response(body)

File: typescript/src/discover.ts
import type { Transport } from "./transport.js";
import type {
  DiscoverArgs,
  DiscoveredService,
  ServiceEndpoint,
  ServicePaymentMethod,
} from "./types.js";

interface DiscoverResponse {
  services: Array<{
    slug: string;
    name: string;
    description: string | null;
    category: string;
    logoUrl: string | null;
    docsUrl: string | null;
    paymentMethods: ServicePaymentMethod[];
    endpoints: ServiceEndpoint[];
    preferredBaseUrl: string | null;
    policyAllowed: boolean;
  }>;
  count: number;
}

/**
 * Calls `GET /api/services` and returns the parsed list. The caller (the
 * `Canopy` client) supplies `agent_id` so the backend can apply policy-aware
 * filtering: services not on the agent's slug allowlist are filtered out by
 * default (or marked `policyAllowed: false` when `includeBlocked: true`).
 *
 * Each service's `preferredBaseUrl` is picked by treasury funding — the rail
 * whose chain has positive USDC. Use `fetch(preferredBaseUrl + endpoint.path)`
 * to call a service.
 */
export async function discover(
  transport: Transport,
  agentId: string | undefined,
  args: DiscoverArgs = {},
): Promise<DiscoveredService[]> {
  const params = new URLSearchParams();
  if (args.category) {
    const cats = Array.isArray(args.category) ? args.category : [args.category];
    for (const c of cats) params.append("category", c);
  }
  if (args.query) params.set("q", args.query);
  if (args.includeUnverified) params.set("include_unverified", "true");
  if (args.includeBlocked) params.set("include_blocked", "true");
  if (args.limit !== undefined) params.set("limit", String(args.limit));
  if (agentId) params.set("agent_id", agentId);

  const path = `/api/services${params.toString() ? `?${params.toString()}` : ""}`;
  const { body } = await transport.request<DiscoverResponse>({
    method: "GET",
    path,
    expectStatuses: [200],
  });

  return body.services.map((s) => ({
    slug: s.slug,
    name: s.name,
    description: s.description,
    category: s.category,
    logoUrl: s.logoUrl,
    docsUrl: s.docsUrl,
    paymentMethods: s.paymentMethods,
    endpoints: s.endpoints,
    preferredBaseUrl: s.preferredBaseUrl,
    policyAllowed: s.policyAllowed,
  }));
}
